zebra-state/src/service/finalized_state/zebra_db/shielded.rs

```diff
@@ -289,6 +290,33 @@ impl ZebraDb {
         }
     }
 
+    /// Get the sapling note commitment subtress for the finalized tip.
+    #[allow(clippy::unwrap_in_result)]
+    fn sapling_subtree_for_tip(&self) -> Option<NoteCommitmentSubtree<sapling::tree::Node>> {
+        let tip_height = self.finalized_tip_height().unwrap_or(Height(0));
+        let sapling_activation_height = NetworkUpgrade::Sapling
+            .activation_height(self.network())
+            .expect("sapling activation height must exist");
+
+        if tip_height >= sapling_activation_height {
+            let sapling_nct_handle = self.db.cf_handle("sapling_note_commitment_tree").unwrap();
+
+            let tip_note_commitment_tree: Arc<sapling::tree::NoteCommitmentTree> = self
+                .db
+                .zs_last_key_value(&sapling_nct_handle)
+                .map(|(_key, value): (RawBytes, _)| Arc::new(value))
+                .expect("Sapling note commitment tree must exist if there is a finalized tip");
+
+            if let Some((subtree_index, node)) =
+                tip_note_commitment_tree.completed_subtree_index_and_root()
+            {
+                return Some(NoteCommitmentSubtree::new(subtree_index, tip_height, node));
+            }
+        }
+
+        None
```

Comment:
This should work:

```suggestion
        let sapling_subtrees = self
            .db
            .cf_handle("sapling_note_commitment_subtree")
            .unwrap();

        let (index, subtree_data): (
            NoteCommitmentSubtreeIndex,
            NoteCommitmentSubtreeData<sapling::tree::Node>,
        ) = self.db.zs_last_key_value(&sapling_subtrees)?;

        Some(subtree_data.with_index(index))
```

Reply:
I think this code is functional, but doesn't actually do what we want (or what this method is documented to do):
https://github.com/ZcashFoundation/zebra/pull/7636/files#r1340708466

Reply:
It could filter the return value unless the end height matches the finalized tip height?

Reply:
Code suggestion was added in https://github.com/ZcashFoundation/zebra/pull/7636/commits/84e79009ecb7ffbc41e056278d99151018070430
Filter was added in https://github.com/ZcashFoundation/zebra/pull/7636/commits/5866bbba3f9b4e025c91b6c150c10e9ff937d494